File: Trees/BinaryTrees/LCA.cpp
#include <bits/stdc++.h>
using namespace std;

typedef struct TreeNode{
    int val;
    TreeNode* left,*right;

    TreeNode(){
    }

    TreeNode(int val){
        this->val = val;
        left = right = NULL;
    }
}TreeNode;

class Util{
	public:
    static TreeNode* getTree(){
        TreeNode* root = new TreeNode(10);
        root->left = new TreeNode(20);
        root->right = new TreeNode(30);

        root->left->left = new TreeNode(40);
        root->left->right = new TreeNode(50);
        root->left->right->left = new TreeNode(60);
        root->left->right->right = new TreeNode(70);
        return root;
    }
    static int LCA(TreeNode* root,int targetNode1,int targetNode2){
        if(root == nullptr) return -1;

        if(root->val == targetNode1 || root->val == targetNode2)    return root->val;

        int left = LCA(root->left,targetNode1,targetNode2);
        int right = LCA(root->right,targetNode1,targetNode2);

        if((left == targetNode1 && right == targetNode2) || (left == targetNode2 && right == targetNode1))  return root->val;

        return left != -1 ? left : right;
    }
};
int main(){
   return 0;
}

File: Trees/BinaryTrees/LCA.java
package Trees.BinaryTrees;

public class LCA {

    private static TreeNode getLCA(TreeNode root, int target1, int target2) {
        if (root == null) return null;

        if (root.data == target1 || root.data == target2) return root;

        TreeNode left = getLCA(root.left, target1, target2);
        TreeNode right = getLCA(root.right, target1, target2);
        if (left != null && right != null)
            if (left.data == target1 && right.data == target2 || left.data == target2 && right.data == target1)
                return root;
        return left != null ? left : right;
    }

    static int[] level = new int[10];
    static int[][] p = new int[30][30];

    private static void dfs(int src, int par, int lvl, int[][] graph) {
        level[src] = lvl;
        for (int child : graph[src]) {
            if (child != par) {
                dfs(child, src, lvl + 1, graph);
            }
        }
    }

    public static void main(String[] args) {
        TreeNode root = TreeNode.getInstance1();
        TreeNode ans = getLCA(root, 3, 8);
        System.out.println(ans);
    }
}

